groups/ntc/ntcu/ntcu_timestampcorrelator.h

```diff
@@ -0,0 +1,160 @@
+/// ids.  Timestamp ids are provided externally
+///
+/// @par Thread Safety
+/// This class is not thread safe.
+///
+/// @ingroup module_ntcu
+class TimestampCorrelator
+{
+  public:
+    typedef bsls::TimeInterval       Timestamp;
+    typedef bsl::uint32_t            ID;
+    typedef bsl::pair<ID, Timestamp> IdTimestamp;
+
+  private:
+    enum { k_RING_BUFFER_SIZE = 64 };
+
+    /// @internal @brief
+    /// RingBuffer is used to store items and extract them based on their id.
+
+    class RingBuffer
+    {
+      public:
+        typedef bsl::uint16_t size_type;
+        struct Data {
+            bool      free;
+            ID        id;
+            Timestamp ts;
+        };
+
+        ///  Construct ring buffer using specified 'capacity' and 'alloc'.
+        /// 'capacity' must be a power of 2.
+        RingBuffer(size_type capacity, bslma::Allocator* alloc);
+
+        /// Save specified 'ts' with specified 'id' into the buffer
+        void push(ID id, const Timestamp& ts);
+
+        /// Search for timestamp with specified 'id'. If such timestamp exists
+        /// then return it. Otherwise return an empty value.
+        bdlb::NullableValue<Timestamp> extract(ID id);
+
+        /// Do reset internal buffer and get ready to work from scratch
+        void reset();
+
+        /// Return wrap-around index after the specified 'val'
+        size_type nextIndex(size_type val) const;
+
+        /// Return string representation of the specified buffer
+        bsl::string toString() const;
+
+        /// Defines the traits of this type. These traits can be used to select,
+        /// at compile-time, the most efficient algorithm to manipulate objects
+        /// of this type.
+        NTSCFG_DECLARE_NESTED_USES_ALLOCATOR_TRAITS(RingBuffer);
+
+      private:
+        size_type         d_currentSize;
+        size_type         d_nextPush;
+        bsl::vector<Data> d_buffer;
+    };
+
+  public:
+    /// Construct TimestampCorrelator using the specified 'alloc. Use the
+    /// specified 'streamProto' set to true to indicate that stream protocol
+    /// is used. Otherwise set it to false.
+    TimestampCorrelator(bool streamProto, bslma::Allocator* alloc);
+
+    /// Destructor
+    ~TimestampCorrelator();
+
+    /// Save the specified 'ts' timestamp into the ring buffer. Use the
+    /// specified 'id' as its identifier.
+    void saveTimestampBeforeSend(const bsls::TimeInterval& ts, ID id);
+
+    /// Search for timestamp which corresponds to specified 'ots'. If found,
+    /// return its value. Otherwise return an empty value.
+    bdlb::NullableValue<bsls::TimeInterval> timestampReceived(
+        const ntsa::Timestamp& ts);
+
+    /// Do reset operation and get ready to work from scratch
+    void reset();
+
+    /// Function used for testing purposes only. Returns number of elements
+    /// of an internal RingBuffer
+    BSLS_KEYWORD_CONSTEXPR static bsl::size_t getRbSize();
```

Comment:
Make `k_RING_BUFFER_SIZE` public and remove this function.

Reply:
Done